airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/jsonschema/JsonSchemaPreprocessor.kt

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.integrations.destination.s3.jsonschema
+
+import com.fasterxml.jackson.databind.node.ObjectNode
+import io.airbyte.commons.jackson.MoreMappers
+
+class JsonSchemaPreprocessor : JsonSchemaIdentityMapper() {
+    /*
+       This is a kludge to move "merging unions of objects" upstream
+       of the AvroPreprocessor.
+
+       The correct place for this in the new CDK is in the process
+       of building the coming bespoke AirbyteSchema, which should
+       normalize the types into a single non-json representation.
+    */
+
+    // Merge the right object's properties into the left's, recursively
+    private fun mergeRightObjectIntoLeft(left: ObjectNode, right: ObjectNode) {
+        val leftProperties = left["properties"] as ObjectNode
+        val rightProperties = right["properties"] as ObjectNode
+        rightProperties.fields().forEach { (rightPropertyName, rightPropertySchema) ->
+            if (!leftProperties.has(rightPropertyName)) {
+                // Just add the property if it doesn't exist
+                leftProperties.set<ObjectNode>(rightPropertyName, rightPropertySchema as ObjectNode)
+            } else {
+                // Otherwise recursively merge
+                val leftPropertySchema = leftProperties[rightPropertyName] as ObjectNode
+                if (leftPropertySchema != rightPropertySchema) {
+                    val leftType = AirbyteJsonSchemaType.fromJsonSchema(leftPropertySchema)
+                    val rightType =
+                        AirbyteJsonSchemaType.fromJsonSchema(rightPropertySchema as ObjectNode)
+                    if (
+                        leftType == AirbyteJsonSchemaType.OBJECT_WITH_PROPERTIES &&
+                            rightType == leftType
+                    ) {
+                        mergeRightObjectIntoLeft(leftPropertySchema, rightPropertySchema)
+                    } else {
+                        throw IllegalArgumentException(
+                            "Cannot merge properties of different types ($rightPropertyName: $leftType, $rightType)"
+                        )
+                    }
+                } // Do nothing if the properties are the same
+            }
+        }
+    }
+
+    override fun mapUnion(schema: ObjectNode): ObjectNode {
+        val newSchema = MoreMappers.initMapper().createObjectNode()
+        val newOptions = MoreMappers.initMapper().createArrayNode()
+
+        // Start an empty mergeable object
+        val mergedObj = MoreMappers.initMapper().createObjectNode()
+        mergedObj.put("type", "object")
+        val mergedProperties = MoreMappers.initMapper().createObjectNode()
+        mergedObj.replace("properties", mergedProperties)
+
+        val seenSet = mutableSetOf<ObjectNode>()
+        for (oldOption in schema["oneOf"]) {
```

Comment:
for my understanding:
1. if we have `oneOf: [{type: object}, {type: object, properties: {...}}]`, this currently leaves both of those objects untouched? (did I read this correctly?)
2. is that correct behavior? (does the later transformation of `{type: object}` to `{type: string}` solve this?)

(... there's something interesting here, that I'm already losing track of what mappers exist + how they interact :/ not sure how to deal with that)

Reply:
Yes, the object without properties and the object with properties are different and won't be merged.

The object without properties will become a string type, so the resulting union would be `["null", "string", /* record */ ]`

Part of the problem comes from the fact that the protocol itself is confusing. I think I can address some of that when finalizing this for the CDK by

* naming stuff after exactly what it does
* identifying everything in the domain of "rationalize maybe bad schemas" (like dealing with union ambiguities) and doing it first generically

Reply:
This is the confusing part: we have three features that combined create a circular dependency:

* converting some types into other types
* merging like unioned types (with the possible result of "no longer a union")
* promoting unions to disjoint records

Reply:
Just to be safe, I forced `"properties": null` and `"properties": {}` to `OBJECT_WITHOUT_PROPERTIES` and added some more test coverage. (The latter case is actually one that AJ flagged: a lot of upstream sources don't report their schemas, so clients just end up with a massive set of _additional_properties. But actually pure empty records can break parquet.

Reply:
(I'll also add clarification around this to the docs.)